File: frontend/types/complaint/complaint.ts
/**
 * `complaints` entity — mirrors the `complaints` table in the schema diagram.
 */

export type ComplaintStatus =
  | "Submitted"
  | "Assigned"
  | "UnderReview"
  | "ReportSubmitted"
  | "AwaitingBudget"
  | "BudgetAllocated"
  | "TenderNotificationIssued"
  | "TenderAllotted"
  | "WorkInProgress"
  | "WorkCompleted"
  | "Resolved"
  | "Reopened"
  | "Closed"

export type ComplaintPriority = "Low" | "Medium" | "High" | "Critical"

/** How staff settled a citizen's dispute over an issue link. */
export type DisputeOutcome = "Upheld" | "Rejected"

/** An officer's hand-back of a complaint, shown to whoever re-allots it. */
export interface AssignmentRejection {
  officerId: string
  officerName: string | null
  reason: string | null
  rejectedAt: string | null
}

export const COMPLAINT_STATUSES: readonly ComplaintStatus[] = [
  "Submitted",
  "Assigned",
  "UnderReview",
  "ReportSubmitted",
  "AwaitingBudget",
  "BudgetAllocated",
  "TenderNotificationIssued",
  "TenderAllotted",
  "WorkInProgress",
  "WorkCompleted",
  "Resolved",
  "Reopened",
  "Closed",
]

export const COMPLAINT_PRIORITIES: readonly ComplaintPriority[] = [
  "Low",
  "Medium",
  "High",
  "Critical",
]

export interface Complaint {
  id: string // uuid
  title: string
  description: string
  citizenId: string // FK to users.id
  departmentId: number // FK to departments.id
  /**
   * The officer (users.id / officers.userId) this complaint is allotted to, or
   * `null` while it waits in the department's unassigned queue.
   */
  assignedOfficerId: string | null
  priority: ComplaintPriority
  status: ComplaintStatus
  latitude: number | null
  longitude: number | null
  address: string
  locality: string
  city: string
  district: string
  state: string
  pincode: string
  country: string
  aiCategory: string | null
  aiPriorityScore: number | null
  clusterId: string | null
  isClusterPrimary: boolean
  clusterDisputed: boolean
  clusterReportCount: number
  clusterPrimaryId: string | null
  clusterPrimaryTitle: string | null
  /** Why the reporter contested the link; null unless a dispute was raised. */
  disputeReason: string | null
  disputeRaisedAt: string | null
  /** How staff settled it; null while a dispute is open or was never raised. */
  disputeOutcome: DisputeOutcome | null
  disputeResolutionNote: string | null
  disputeResolvedAt: string | null
  disputeResolvedByName: string | null
  /** Officers who already rejected this complaint, newest first. */
  rejectionHistory: AssignmentRejection[]
  /** Budget committed to this complaint, set on allocation; null until then. */
  allocatedBudget: number | null
  /** Financial year the budget was drawn from (e.g. "2026-27"). */
  budgetYear: string | null
  createdAt: string
  updatedAt: string
}

File: backend/app/models/complaint.py
import uuid

from app.extensions import db
from app.models.base_model import BaseModel
from app.models.enums import ComplaintPriority, ComplaintStatus


class Complaint(BaseModel):
    __tablename__ = "complaints"

    id = db.Column(
        db.UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
    )

    title = db.Column(
        db.String(255),
        nullable=False,
    )

    description = db.Column(
        db.Text,
        nullable=False,
    )

    citizen_id = db.Column(
        db.UUID(as_uuid=True),
        db.ForeignKey("users.id"),
        nullable=False,
    )

    department_id = db.Column(
        db.Integer,
        db.ForeignKey("departments.id"),
        nullable=False,
    )

    priority = db.Column(
        db.Enum(ComplaintPriority),
        nullable=False,
    )

    status = db.Column(
        db.Enum(ComplaintStatus),
        default=ComplaintStatus.SUBMITTED,
        nullable=False,
    )

    latitude = db.Column(
        db.Numeric(10, 7),
        nullable=True,
    )

    longitude = db.Column(
        db.Numeric(10, 7),
        nullable=True,
    )

    address = db.Column(
        db.Text,
        nullable=False,
    )

    locality = db.Column(
        db.Text,
        nullable=False,
    )

    city = db.Column(
        db.String(100),
        nullable=False,
    )

    district = db.Column(
        db.String(100),
        nullable=True,
    )

    state = db.Column(
        db.String(100),
        nullable=False,
    )

    country = db.Column(
        db.String(100),
        nullable=True,
    )

    pincode = db.Column(
        db.String(10),
        nullable=False,
    )

    ai_category = db.Column(
        db.String(100),
        nullable=False,
    )

    ai_priority_score = db.Column(
        db.Integer,
        nullable=False,
    )

    # -------------------------
    # Relationships
    # -------------------------

    citizen = db.relationship(
        "User",
        back_populates="complaints",
    )

    department = db.relationship(
        "Department",
        back_populates="complaints",
    )

    images = db.relationship(
        "ComplaintImage",
        back_populates="complaint",
        cascade="all, delete-orphan",
    )

    remarks = db.relationship(
        "ComplaintRemark",
        back_populates="complaint",
        cascade="all, delete-orphan",
    )

    assignments = db.relationship(
        "ComplaintAssignment",
        back_populates="complaint",
        cascade="all, delete-orphan",
    )

    review_report = db.relationship(
        "ReviewReport",
        back_populates="complaint",
        uselist=False,
        cascade="all, delete-orphan",
    )

    tender = db.relationship(
        "Tender",
        back_populates="complaint",
        uselist=False,
        cascade="all, delete-orphan",
    )

    def __repr__(self):
        return f"<Complaint {self.id}>"
